File: implementations/cpp/treap.cpp
#include <iostream>
#include <vector>
#include <random>
#include <cassert>
#include <stdexcept>
#include <optional>

namespace dsa {

/**
 * @brief Randomized Cartesian Tree (Treap).
 * Combines Binary Search Tree on keys with Max-Heap on random priorities.
 * Provides expected O(log n) search, insert, and delete using split and merge.
 */
class Treap {
public:
    struct Node {
        int key;
        uint32_t priority;
        size_t size;
        Node* left;
        Node* right;

        Node(int k, uint32_t p)
            : key(k), priority(p), size(1), left(nullptr), right(nullptr) {}
    };

    Treap() : root_(nullptr), rng_(1337) {}

    explicit Treap(uint32_t seed) : root_(nullptr), rng_(seed) {}

    ~Treap() {
        clear(root_);
    }

    Treap(const Treap&) = delete;
    Treap& operator=(const Treap&) = delete;

    bool empty() const { return root_ == nullptr; }
    size_t size() const { return get_size(root_); }

    /**
     * @brief Inserts a key. If duplicate, returns false.
     */
    bool insert(int key) {
        if (contains(key)) return false;

        uint32_t p = rng_();
        Node* new_node = new Node(key, p);

        Node* left = nullptr;
        Node* right = nullptr;
        split(root_, key, left, right);

        root_ = merge(merge(left, new_node), right);
        return true;
    }

    /**
     * @brief Erases a key. Returns true if removed, false if not found.
     */
    bool erase(int key) {
        if (!contains(key)) return false;

        Node* left = nullptr;
        Node* mid = nullptr;
        Node* right = nullptr;

        // Split into < key and >= key
        split(root_, key - 1, left, right);
        // Split right into == key and > key
        split(right, key, mid, right);

        // Delete mid node
        delete mid;

        // Merge left and right back
        root_ = merge(left, right);
        return true;
    }

    /**
     * @brief Checks if a key exists in the Treap.
     */
    bool contains(int key) const {
        Node* cur = root_;
        while (cur) {
            if (cur->key == key) return true;
            cur = (key < cur->key) ? cur->left : cur->right;
        }
        return false;
    }

    /**
     * @brief Finds the k-th smallest element (0-indexed).
     */
    std::optional<int> kth_element(size_t k) const {
        if (k >= size()) return std::nullopt;
        Node* cur = root_;
        while (cur) {
            size_t left_sz = get_size(cur->left);
            if (k == left_sz) {
                return cur->key;
            } else if (k < left_sz) {
                cur = cur->left;
            } else {
                k -= (left_sz + 1);
                cur = cur->right;
            }
        }
        return std::nullopt;
    }

    /**
     * @brief Returns sorted keys via in-order traversal.
     */
    std::vector<int> inorder() const {
        std::vector<int> result;
        inorder_recursive(root_, result);
        return result;
    }

    /**
     * @brief Verifies that both BST and Heap invariants hold throughout the tree.
     */
    bool verify_invariants() const {
        return verify_node(root_, nullptr, nullptr);
    }

private:
    Node* root_;
    std::mt19937 rng_;

    static size_t get_size(Node* node) {
        return node ? node->size : 0;
    }

    static void update_size(Node* node) {
        if (node) {
            node->size = 1 + get_size(node->left) + get_size(node->right);
        }
    }

    void clear(Node* node) {
        if (!node) return;
        clear(node->left);
        clear(node->right);
        delete node;
    }

    /**
     * @brief Splits tree rooted at node into left (keys <= key) and right (keys > key).
     */
    void split(Node* node, int key, Node*& left, Node*& right) {
        if (!node) {
            left = nullptr;
            right = nullptr;
            return;
        }

        if (node->key <= key) {
            split(node->right, key, node->right, right);
            left = node;
        } else {
            split(node->left, key, left, node->left);
            right = node;
        }
        update_size(node);
    }

    /**
     * @brief Merges two treaps where all keys in left < all keys in right.
     */
    Node* merge(Node* left, Node* right) {
        if (!left) return right;
        if (!right) return left;

        if (left->priority >= right->priority) {
            left->right = merge(left->right, right);
            update_size(left);
            return left;
        } else {
            right->left = merge(left, right->left);
            update_size(right);
            return right;
        }
    }

    void inorder_recursive(Node* node, std::vector<int>& out) const {
        if (!node) return;
        inorder_recursive(node->left, out);
        out.push_back(node->key);
        inorder_recursive(node->right, out);
    }

    bool verify_node(Node* node, Node* min_node, Node* max_node) const {
        if (!node) return true;

        // BST invariant
        if (min_node && node->key <= min_node->key) return false;
        if (max_node && node->key >= max_node->key) return false;

        // Max-heap invariant
        if (node->left && node->left->priority > node->priority) return false;
        if (node->right && node->right->priority > node->priority) return false;

        // Subtree size check
        if (node->size != 1 + get_size(node->left) + get_size(node->right)) return false;

        return verify_node(node->left, min_node, node) &&
               verify_node(node->right, node, max_node);
    }
};

} // namespace dsa

// ============================================================================
// Unit Tests
// ============================================================================

void test_treap_basic() {
    dsa::Treap tr(42);
    assert(tr.empty());
    assert(tr.size() == 0);

    std::vector<int> keys = {50, 30, 70, 20, 40, 60, 80};
    for (int k : keys) {
        assert(tr.insert(k));
    }
    assert(tr.size() == 7);
    assert(!tr.insert(50)); // Duplicate rejected

    assert(tr.verify_invariants());

    // Search
    for (int k : keys) {
        assert(tr.contains(k));
    }
    assert(!tr.contains(999));
    assert(!tr.contains(10));

    // In-order traversal
    std::vector<int> expected = {20, 30, 40, 50, 60, 70, 80};
    assert(tr.inorder() == expected);

    // K-th element (0-indexed)
    assert(tr.kth_element(0).value() == 20);
    assert(tr.kth_element(3).value() == 50);
    assert(tr.kth_element(6).value() == 80);
    assert(tr.kth_element(7) == std::nullopt);
}

void test_treap_erase() {
    dsa::Treap tr(42);
    for (int k : {50, 30, 70, 20, 40, 60, 80}) {
        tr.insert(k);
    }

    assert(tr.erase(20));
    assert(!tr.contains(20));
    assert(tr.size() == 6);
    assert(tr.verify_invariants());

    assert(tr.erase(50)); // Root or internal node
    assert(!tr.contains(50));
    assert(tr.size() == 5);
    assert(tr.verify_invariants());

    std::vector<int> expected = {30, 40, 60, 70, 80};
    assert(tr.inorder() == expected);

    assert(!tr.erase(999));
}

int main() {
    std::cout << "Running Treap C++17 unit tests...\n";
    test_treap_basic();
    test_treap_erase();
    std::cout << "All Treap C++17 unit tests passed successfully!\n";
    return 0;
}

File: implementations/python/treap.py
"""
Treap (Randomized Cartesian Tree)

This module provides reference implementations of:
- Treap: Combining BST on keys with Heap on random priorities via split and merge
"""

from typing import List, Optional, Tuple
import random
import unittest


class Treap:
    class _Node:
        def __init__(self, key: int, priority: int):
            self.key = key
            self.priority = priority
            self.size = 1
            self.left: Optional['Treap._Node'] = None
            self.right: Optional['Treap._Node'] = None

    def __init__(self, seed: int = 1337):
        self._root: Optional[Treap._Node] = None
        self._rng = random.Random(seed)

    def empty(self) -> bool:
        return self._root is None

    def size(self) -> int:
        return self._get_size(self._root)

    def _get_size(self, node: Optional['Treap._Node']) -> int:
        return node.size if node else 0

    def _update_size(self, node: Optional['Treap._Node']) -> None:
        if node:
            node.size = 1 + self._get_size(node.left) + self._get_size(node.right)

    def _split(
        self, node: Optional['Treap._Node'], key: int
    ) -> Tuple[Optional['Treap._Node'], Optional['Treap._Node']]:
        """Splits tree rooted at node into left (keys <= key) and right (keys > key)."""
        if not node:
            return None, None

        if node.key <= key:
            node.right, right = self._split(node.right, key)
            self._update_size(node)
            return node, right
        else:
            left, node.left = self._split(node.left, key)
            self._update_size(node)
            return left, node

    def _merge(
        self, left: Optional['Treap._Node'], right: Optional['Treap._Node']
    ) -> Optional['Treap._Node']:
        """Merges two treaps where all keys in left < all keys in right."""
        if not left:
            return right
        if not right:
            return left

        if left.priority >= right.priority:
            left.right = self._merge(left.right, right)
            self._update_size(left)
            return left
        else:
            right.left = self._merge(left, right.left)
            self._update_size(right)
            return right

    def insert(self, key: int) -> bool:
        """Inserts a key. Returns False if duplicate."""
        if self.contains(key):
            return False

        p = self._rng.randint(0, 1 << 30)
        new_node = self._Node(key, p)

        left, right = self._split(self._root, key)
        self._root = self._merge(self._merge(left, new_node), right)
        return True

    def erase(self, key: int) -> bool:
        """Removes a key. Returns True if removed, False if not found."""
        if not self.contains(key):
            return False

        left, right = self._split(self._root, key - 1)
        mid, right = self._split(right, key)

        self._root = self._merge(left, right)
        return True

    def contains(self, key: int) -> bool:
        """Searches for key following BST property."""
        cur = self._root
        while cur:
            if cur.key == key:
                return True
            cur = cur.left if key < cur.key else cur.right
        return False

    def kth_element(self, k: int) -> Optional[int]:
        """Returns k-th smallest element (0-indexed)."""
        if k < 0 or k >= self.size():
            return None
        cur = self._root
        while cur:
            left_sz = self._get_size(cur.left)
            if k == left_sz:
                return cur.key
            elif k < left_sz:
                cur = cur.left
            else:
                k -= left_sz + 1
                cur = cur.right
        return None

    def inorder(self) -> List[int]:
        """Returns keys in sorted order."""
        res: List[int] = []

        def _inorder(node: Optional['Treap._Node']):
            if not node:
                return
            _inorder(node.left)
            res.append(node.key)
            _inorder(node.right)

        _inorder(self._root)
        return res

    def verify_invariants(self) -> bool:
        def _verify(node, min_k, max_k):
            if not node:
                return True
            if min_k is not None and node.key <= min_k:
                return False
            if max_k is not None and node.key >= max_k:
                return False
            if node.left and node.left.priority > node.priority:
                return False
            if node.right and node.right.priority > node.priority:
                return False
            if node.size != 1 + self._get_size(node.left) + self._get_size(node.right):
                return False
            return _verify(node.left, min_k, node.key) and _verify(node.right, node.key, max_k)

        return _verify(self._root, None, None)


# ============================================================================
# Unit Tests
# ============================================================================

class TestTreap(unittest.TestCase):
    def test_treap_basic(self):
        tr = Treap(42)
        self.assertTrue(tr.empty())
        keys = [50, 30, 70, 20, 40, 60, 80]
        for k in keys:
            self.assertTrue(tr.insert(k))

        self.assertFalse(tr.insert(50))
        self.assertEqual(tr.size(), 7)
        self.assertTrue(tr.verify_invariants())

        for k in keys:
            self.assertTrue(tr.contains(k))
        self.assertFalse(tr.contains(999))

        self.assertEqual(tr.inorder(), [20, 30, 40, 50, 60, 70, 80])
        self.assertEqual(tr.kth_element(0), 20)
        self.assertEqual(tr.kth_element(3), 50)
        self.assertEqual(tr.kth_element(6), 80)
        self.assertIsNone(tr.kth_element(7))

    def test_treap_erase(self):
        tr = Treap(42)
        for k in [50, 30, 70, 20, 40, 60, 80]:
            tr.insert(k)

        self.assertTrue(tr.erase(20))
        self.assertFalse(tr.contains(20))
        self.assertTrue(tr.verify_invariants())

        self.assertTrue(tr.erase(50))
        self.assertFalse(tr.contains(50))
        self.assertTrue(tr.verify_invariants())

        self.assertEqual(tr.inorder(), [30, 40, 60, 70, 80])
        self.assertFalse(tr.erase(999))


if __name__ == "__main__":
    unittest.main()
